photon-core/src/main/java/org/photonvision/vision/pipeline/ArucoPipeline.java

```diff
@@ -69,59 +76,171 @@ public ArucoPipeline(ArucoPipelineSettings settings) {
+        tagDetectionPipeResult = arucoDetectionPipe.run(frame.processedImage);
+        sumPipeNanosElapsed += tagDetectionPipeResult.nanosElapsed;
+
+        // If we want to debug the thresholding steps, draw the first step to the color image
+        if (settings.debugThreshold) {
+            drawThresholdFrame(
+                    frame.processedImage.getMat(),
+                    frame.colorImage.getMat(),
+                    settings.threshWinSizes.getFirst(),
+                    settings.threshConstant);
         }
 
-        tagDetectionPipeResult = arucoDetectionPipe.run(rawInputMat);
-        targetList = new ArrayList<>();
+        List<TrackedTarget> targetList = new ArrayList<>();
         for (ArucoDetectionResult detection : tagDetectionPipeResult.output) {
-            // TODO this should be in a pipe, not in the top level here (Matt)
-
-            // populate the target list
-            // Challenge here is that TrackedTarget functions with OpenCV Contour
+            // Populate target list for multitag
+            // (TODO: Address circular dependencies. Multitag only requires corners and IDs, this should
+            // not be necessary.)
             TrackedTarget target =
                     new TrackedTarget(
                             detection,
+                            null,
                             new TargetCalculationParameters(
                                     false, null, null, null, null, frameStaticProperties));
 
-            var correctedBestPose = target.getBestCameraToTarget3d();
+            targetList.add(target);
+        }
 
-            target.setBestCameraToTarget3d(
-                    new Transform3d(correctedBestPose.getTranslation(), correctedBestPose.getRotation()));
+        // Do multi-tag pose estimation
+        MultiTargetPNPResults multiTagResult = new MultiTargetPNPResults();
+        if (settings.solvePNPEnabled && settings.doMultiTarget) {
+            var multiTagOutput = multiTagPNPPipe.run(targetList);
+            sumPipeNanosElapsed += multiTagOutput.nanosElapsed;
+            multiTagResult = multiTagOutput.output;
+        }
 
-            targetList.add(target);
+        // Do single-tag pose estimation
+        if (settings.solvePNPEnabled) {
+            // Clear target list that was used for multitag so we can add target transforms
+            targetList.clear();
+            // TODO global state again ew
+            var atfl = ConfigManager.getInstance().getConfig().getApriltagFieldLayout();
+
+            for (ArucoDetectionResult detection : tagDetectionPipeResult.output) {
+                AprilTagPoseEstimate tagPoseEstimate = null;
+                // Do single-tag estimation when "always enabled" or if a tag was not used for multitag
+                if (settings.doSingleTargetAlways
+                        || !multiTagResult.fiducialIDsUsed.contains(Integer.valueOf(detection.getId()))) {
```

Comment:
Nit: `ArucoDetectionResult::getId` already returns an int so this is redundant boxing